gitclone/steperror.go

```diff
@@ -1,11 +1,178 @@
 package gitclone
 
-import "github.com/bitrise-io/bitrise-init/step"
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/bitrise-io/bitrise-init/step"
+)
+
+// DetailBuilder ...
+type DetailBuilder = func(...string) Detail
+
+// PatternErrorMatcher ...
+type PatternErrorMatcher struct {
+	defaultHandler DetailBuilder
+	handlers       map[string]DetailBuilder
+}
+
+func newPatternErrorMatcher(defaultHandler DetailBuilder, handlers map[string]DetailBuilder) *PatternErrorMatcher {
+	m := PatternErrorMatcher{
+		handlers:       handlers,
+		defaultHandler: defaultHandler,
+	}
+
+	return &m
+}
+
+// Run ...
+func (m *PatternErrorMatcher) Run(msg string) step.Recommendation {
+	for pattern, handler := range m.handlers {
+		re := regexp.MustCompile(pattern)
+		if re.MatchString(msg) {
+			matches := re.FindStringSubmatch((msg))
+
+			if len(matches) > 1 {
```

Comment:
i suggest introducing a new variable to store the match groups, this way we express that the `handler` expects the match groups (and not the first item of `matches`: the whole string matches to the pattern) as input.

by introducing the match groups variable, we can express here, that some handler receives nil/empty slice

Reply:
Yep, good idea.
Since we are having at least [search_string], and we left off the first param, we will end up with an empty slice if there is no match group defined in the regex.
`matches[1:]` returns either `[]` or [`match1, match2, ...]`